Update point_is_in_domain function

anne-glerum opened the pull request:
We have use cases where we apply traction boundary conditions and mesh deformation. However, the initial lithostatic pressure traction plugin uses the function `point_is_in_domain` of the geometry model to check whether the reference point where the lithostatic profile is calculated lies in the domain. At the moment, the function cannot deal with initial topography and throws an error message. It also neglects initial mesh deformation. However, for `chunk` and `box` geometries, we know what initial topography is applied. And we can also check whether initial mesh deformation is applied. This PR:
* includes max initial topography at the queried point in the domain extents for box and chunk geometries
* checks whether mesh deformation is applied in the box model. If so, it will throw when initial mesh deformation is applied, or if time-dependent mesh deformation is applied after timestep 0

The problem is that when the lithostatic profile is computed, initialization of the simulator has not finished, and therefore we cannot ask for the Mesh Deformation Handler. My original problem is therefore not solved. I do consider this PR an improvement by itself and would appreciate input on how to also fix my problem. I could read in the required mesh deformation information from the input file, but that seems less elegant.

The `initial lithostatic pressure` plugin actually does not need to know whether the representative point lies within the domain boundaries in all directions, only in the horizontal directions, as the vertical coordinate is ignored. The `point_is_in_domain` function could therefore become a `point_is_within_horizontal_extents_of_domain` function. 
But, the `point_is_in_domain` function is by now used in one other place than for the `initial lithostatic pressure` plugin, namely in initial_temperature/random_gaussian_perturbation.cc. Therefore, I could also do the checking of the horizontal extents within the `initial lithostatic pressure` plugin, as this plugin requests a lot of information from the geometry models already.



### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

anne-glerum commented:
Ah good idea! I'll have a look if using find_active_cell_around_point can cover all the possible scenarios, thanks!

gassmoeller commented:
I took a closer look at the code, but I have to say it is quite hard to understand its intent. I have two question:
- for the box geometry: is the only change in this PR that now `point_is_in_domain` can be used in timestep 0 (after simulator initialization) with mesh deformation unless the `ascii data` mesh deformation is active?
- for the chunk geometry: Why can you remove the assert there, and not add the same conditions as for the box?

If the solution to expand the functionality of the functions to also check deformed meshes is working, I think that would be what I prefer to the current approach in this PR.

anne-glerum commented:
I've now also added the find_active_cell_around_point call to chunk.cc. Two new tests check whether the point_is_in_domain function works with initial topography for the box and the chunk. And I've adapted an existing test to test the function over time with mesh deformation. All tests fail on my laptop though, so I have to see what the tester says.

bangerth commented on source/geometry_model/box.cc:
```diff
@@ -316,16 +317,60 @@ namespace aspect
-      AssertThrow(!this->get_parameters().mesh_deformation_enabled ||
-                  this->simulator_is_past_initialization() == false,
-                  ExcMessage("After displacement of the free surface, this function can no longer be used to determine whether a point lies in the domain or not."));
+      // If mesh deformation occurs over time, and we're past the
+      // first timestep (during which no mesh deformation is applied),
+      // we can no longer use the original extents of the model domain
+      // to check whether the point lies in the domain.
+      // Initial mesh deformation is applied during the first
+      // timestep, but this function might be called before the
+      // simulator is initialized and thus before we can ask whether initial
+      // or time-dependent mesh deformation is applied.
+      // Therefore, when any mesh deformation is enabled,
+      // we loop over all the current grid cells to see if the point
+      // lies within the domain.
+      // Do note that this function can be called before mesh deformation
+      // is applied in the first timestep, and therefore there is no guarantee
+      // that the point will lie in the domain after initial mesh deformation.
+      // For example, boundary traction plugins are initialized before
+      // mesh deformation plugins and the initial_lithostatic_pressure
+      // plugin calls this function while the simulator is being initialized.
+      // However, if the simulator has not been initialized,
+      // no mesh deformation has been applied and we can use the geometry model's
+      // extents to check whether a point lies in the domain.
+      if (this->get_parameters().mesh_deformation_enabled &&
+          this->simulator_is_past_initialization())
+        {
+          std::pair<const typename parallel::distributed::Triangulation<dim>::active_cell_iterator,
+              Point<dim>> it =
+                GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), point);
```
I am not sure this is a great idea. First, it is quite expensive, though perhaps that is not so important because the function is not called very often. But more importantly, this creates difficulties in parallel computations. There, the mesh consists of locally owned, ghost, and artificial cells, and these differ between processes in both where they are and in their geometric shape. So one process may be returning `true` from this function when another one is returning `false`. That's awkward.

I thought about a solution, and all I can come up with is if all processes call this function, and then they figure whether at least *one* process had a locally owned cell where the point was found. This is then independent of the shapes of ghost and artificial cells. Of course, this is *even more expensive*, and furthermore it makes this function a collective function that needs to be called from all processes at the same time, and presumably with the same points.

gassmoeller replied:
> I thought about a solution, and all I can come up with is if all processes call this function, and then they figure whether at least one process had a locally owned cell where the point was found.

Yes, I think this is the only possible solution. We do something similar in source/postprocess/point_values.cc:75-91. There, we use the function `VectorTools::point_value`, but internally that function uses `find_active_cell_around_point` as well. then the result of the function is evaluated and communicated to all processes.

bangerth commented on source/geometry_model/box.cc:
```diff
@@ -316,16 +317,60 @@ namespace aspect
+            return false;
+        }
+
+      // The maximal extents of the unperturbed box domain.
```
Could you put the rest of this function into an `else` block to make it easier to see that this code is only executed if the `if` above was not entered?

bangerth reviewed (commented):
I am not sure if I like the call to the `GridTools` function.

anne-glerum commented:
Ah yes, of course. I've done something similar to point_values.cc now, is that what you're looking for?

anne-glerum commented:
The new tests run now (but fail as I haven't copied over the test results yet). dealii-master seems broken, 18 unrelated gmg tests fail.

gassmoeller reviewed (commented):
Yes, I think this is the right approach, thanks for the changes. Just some smallish comments. Also please add a changelog, its worth noting that this works now (at least for some geometries) with mesh deformation.

anne-glerum commented:
Hi @gassmoeller, I've added a changelog, unified the comments and moved part of the functionality into a Utilities function. Anything else? The failing tester is only for gmg tests on dealii main.

gassmoeller reviewed (approved):
Thank you. Looks good to go now. :+1:
